src/inmanta/db/versions/v4.py

```diff
@@ -17,11 +17,75 @@
 """
 from asyncpg import Connection
 
+from inmanta.resources import Id
+
 
 async def update(connection: Connection) -> None:
-    schema = """
-    DROP TABLE IF EXISTS public.formrecord;
-    DROP TABLE IF EXISTS public.form;
-    """
     async with connection.transaction():
-        await connection.execute(schema)
+        await connection.execute(
+            """
+--------------------------------
+-- Remove forms functionality --
+--------------------------------
+
+DROP TABLE IF EXISTS public.formrecord;
+DROP TABLE IF EXISTS public.form;
+
+--------------------------------------------------------------------------------
+-- Remove join table resourceversionid and put fields in resourceaction table --
+--------------------------------------------------------------------------------
+
+-- Add new columns to resourceaction table
+ALTER TABLE public.resourceaction
+    ADD COLUMN environment uuid,
+    ADD COLUMN version integer,
+    ADD COLUMN resource_version_ids varchar[];
+
+-- Populate the environment and resource_version_ids columns
+UPDATE public.resourceaction AS r
+SET environment=(SELECT DISTINCT rvi.environment
+                 FROM public.resourceversionid rvi
+                 WHERE rvi.action_id=r.action_id),
+    resource_version_ids=ARRAY(SELECT rvi.resource_version_id
+                               FROM public.resourceversionid rvi
+                               WHERE rvi.action_id=r.action_id);
+
+-- Remove dangling resource actions. Due to a bug, the environment is
+-- unknown when no resources are associated with a resource action.
+DELETE FROM public.resourceaction WHERE environment IS NULL;
+"""
+        )
+
+        # Populate the version column
+        result = await connection.fetch("SELECT action_id, resource_version_ids FROM public.resourceaction")
```

Comment:
this will hammer the database with MANY small queries, we may have to explore alternatives.